File: src/capsolver/api/websocket.py
"""WebSocket live job status updates."""

from __future__ import annotations

import asyncio
import json
from typing import Any

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from capsolver.api.routes import get_manager
from capsolver.jobs.models import Job

router = APIRouter()


@router.websocket("/ws/jobs/{job_id}")
async def job_status_ws(websocket: WebSocket, job_id: str) -> None:
    await websocket.accept()
    manager = get_manager()

    job = await manager.get_job(job_id)
    if not job:
        await websocket.send_json({"error": "Job not found"})
        await websocket.close()
        return

    await websocket.send_json({"type": "snapshot", "job": job.to_public_dict()})

    if job.status.value in ("completed", "failed", "cancelled"):
        await websocket.close()
        return

    update_event = asyncio.Event()
    latest_job: dict[str, Any] = {"job": job}

    async def on_update(updated: Job) -> None:
        if updated.id == job_id:
            latest_job["job"] = updated
            update_event.set()

    manager.subscribe(job_id, on_update)

    try:
        while True:
            try:
                await asyncio.wait_for(update_event.wait(), timeout=30.0)
                update_event.clear()
                current: Job = latest_job["job"]
                await websocket.send_json({"type": "update", "job": current.to_public_dict()})
                if current.status.value in ("completed", "failed", "cancelled"):
                    break
            except asyncio.TimeoutError:
                await websocket.send_json({"type": "ping"})
    except WebSocketDisconnect:
        pass
    finally:
        manager.unsubscribe(job_id, on_update)


@router.websocket("/ws/stats")
async def stats_ws(websocket: WebSocket) -> None:
    await websocket.accept()
    manager = get_manager()

    try:
        while True:
            stats = await manager.get_stats()
            await websocket.send_json({"type": "stats", "data": stats})
            await asyncio.sleep(2)
    except WebSocketDisconnect:
        pass

File: src/api/websocket.js
import { WebSocketServer } from "ws";
import { getManager } from "./routes.js";
import { toPublicJob, isTerminal } from "../jobs/models.js";

export function attachWebSockets(server) {
  const wss = new WebSocketServer({ noServer: true });

  server.on("upgrade", (request, socket, head) => {
    const url = new URL(request.url, `http://${request.headers.host}`);
    const jobMatch = url.pathname.match(/^\/api\/v1\/ws\/jobs\/([^/]+)$/);
    const statsMatch = url.pathname === "/api/v1/ws/stats";

    if (jobMatch) {
      wss.handleUpgrade(request, socket, head, (ws) => {
        handleJobWs(ws, jobMatch[1]);
      });
    } else if (statsMatch) {
      wss.handleUpgrade(request, socket, head, (ws) => {
        handleStatsWs(ws);
      });
    } else {
      socket.destroy();
    }
  });
}

async function handleJobWs(ws, jobId) {
  const manager = getManager();
  const job = await manager.getJob(jobId);

  if (!job) {
    ws.send(JSON.stringify({ error: "Job not found" }));
    ws.close();
    return;
  }

  ws.send(JSON.stringify({ type: "snapshot", job: toPublicJob(job) }));

  if (isTerminal(job.status)) {
    ws.close();
    return;
  }

  let latest = job;
  const onUpdate = async (updated) => {
    if (updated.id !== jobId) return;
    latest = updated;
    ws.send(JSON.stringify({ type: "update", job: toPublicJob(updated) }));
    if (isTerminal(updated.status)) ws.close();
  };

  manager.subscribe(jobId, onUpdate);

  const pingInterval = setInterval(() => {
    if (ws.readyState === ws.OPEN) ws.send(JSON.stringify({ type: "ping" }));
  }, 30000);

  ws.on("close", () => {
    clearInterval(pingInterval);
    manager.unsubscribe(jobId, onUpdate);
  });
}

async function handleStatsWs(ws) {
  const manager = getManager();
  const interval = setInterval(async () => {
    if (ws.readyState !== ws.OPEN) return;
    const stats = await manager.getStats();
    ws.send(JSON.stringify({ type: "stats", data: stats }));
  }, 2000);

  ws.on("close", () => clearInterval(interval));
}
